### PR TITLE
Tests for data sets

### PR DESCRIPTION
**Pull request by palpatel0**
Co-authored by: Tom Fox tom.fox@lis.ac.uk

### CODE REVIEW

File: projtools/tests/test_data.py [hunk 1]
```diff
@@ -73,3 +74,10 @@ def test_borough_multi():
...
+
+def test_data_consistency():
+    """Tests that the data is the same between the different data sets"""
+    assert list(np.sort(data.borough.index.get_level_values(0).unique())) == list(np.sort(data.ward.index.get_level_values(2).unique()))
+    #assert list(np.sort(data.lsoa.index.get_level_values(3).unique())) == list(np.sort(data.ward.index.get_level_values(2).unique())) #LSOA is missing Historical Fraud and Forgery and Sexual Offences; data.lsoa.index.get_level_values(3).unique().append(['Sexual Offences', 'Historical Fraud and Forgery']).sort() doesn't work. Type Error: all inputs must be index
+    assert list(np.sort(data.borough.index.get_level_values(1).unique())) == list(np.sort(data.ward.index.get_level_values(3).unique()))
+    #assert list(np.sort(data.lsoa.index.get_level_values(4).unique())) == list(np.sort(data.ward.index.get_level_values(3).unique())) #LSOA is missing Rape, Other Sexual Offences and Historical Fraud and Forgery
```
Comment (TomWFox) [inline projtools/tests/test_data.py:83]:
```suggestion
def test_data_consistency():
    """Tests that the data is the same between the different data sets"""
    assert list(np.sort(data.borough.index.get_level_values(0).unique())) == list(np.sort(data.ward.index.get_level_values(2).unique()))
    # LSOA is missing Historical Fraud and Forgery and Sexual Offences
    assert list(np.sort(np.append(data.lsoa.index.get_level_values(3).unique().to_numpy(), ['Sexual Offences', 'Historical Fraud and Forgery']))) == list(np.sort(data.ward.index.get_level_values(2).unique()))
    assert list(np.sort(data.borough.index.get_level_values(1).unique())) == list(np.sort(data.ward.index.get_level_values(3).unique()))
    # LSOA is missing Rape, Other Sexual Offences and Historical Fraud and Forgery
    assert list(np.sort(np.append(data.lsoa.index.get_level_values(4).unique().to_numpy(), ['Rape', 'Other Sexual Offences', 'Historical Fraud and Forgery']))) == list(np.sort(data.ward.index.get_level_values(3).unique()))
```

Comment (palpatel0) [reply projtools/tests/test_data.py:83]: how do i edit this code?
I think it would be worth adding comments on what exactly the tests are comparing

Comment (TomWFox) [reply projtools/tests/test_data.py:83]: You can edit in the browser by going to "files changed" and clicking edit

or you can `git switch` your branch to palpatel0:tests and make the change locally, then push

Comment (TomWFox) [reply projtools/tests/test_data.py:83]: but if you mean you want to edit my changes you could first commit, then change whatever you want, or just copy and paste my changes into you local copy, make your changes, and mark this convo as resolved

File: projtools/tests/test_data.py [hunk 2]
```diff
@@ -73,3 +74,12 @@ def test_borough_multi():
...
+
+def test_data_consistency():
+    """Tests that the data is the same between the different data sets"""
+    assert list(np.sort(data.borough.index.get_level_values(0).unique())) == list(np.sort(data.ward.index.get_level_values(2).unique()))
+    # LSOA is missing Historical Fraud and Forgery and Sexual Offences
+    assert list(np.sort(np.append(data.lsoa.index.get_level_values(3).unique().to_numpy(), ['Sexual Offences', 'Historical Fraud and Forgery']))) == list(np.sort(data.ward.index.get_level_values(2).unique()))
+    assert list(np.sort(data.borough.index.get_level_values(1).unique())) == list(np.sort(data.ward.index.get_level_values(3).unique()))
+    # LSOA is missing Rape, Other Sexual Offences and Historical Fraud and Forgery
+    assert list(np.sort(np.append(data.lsoa.index.get_level_values(4).unique().to_numpy(), ['Rape', 'Other Sexual Offences', 'Historical Fraud and Forgery']))) == list(np.sort(data.ward.index.get_level_values(3).unique()))
```
Comment (palpatel0) [inline projtools/tests/test_data.py:85]:
```suggestion
    """Tests that the data is the same between the different data sets"""
    assert list(np.sort(data.borough.index.get_level_values(0).unique())) == list(np.sort(data.ward.index.get_level_values(2).unique())) #This test checks that the borough and ward data sets have the same categories for offences
    # LSOA is missing Historical Fraud and Forgery and Sexual Offences
    assert list(np.sort(np.append(data.lsoa.index.get_level_values(3).unique().to_numpy(), ['Sexual Offences', 'Historical Fraud and Forgery']))) == list(np.sort(data.ward.index.get_level_values(2).unique())) #This test checks that the LSOA and ward datasets have the same categories for offences. Two of the missing categories are added to the LSOA list
    assert list(np.sort(data.borough.index.get_level_values(1).unique())) == list(np.sort(data.ward.index.get_level_values(3).unique())) #This test checks that the borough and ward data sets have the same list of offences
    # LSOA is missing Rape, Other Sexual Offences and Historical Fraud and Forgery
    assert list(np.sort(np.append(data.lsoa.index.get_level_values(4).unique().to_numpy(), ['Rape', 'Other Sexual Offences', 'Historical Fraud and Forgery']))) == list(np.sort(data.ward.index.get_level_values(3).unique())) #This test checks that the LSOA and ward datasets have the same list of offences. Three of the missing offences are added to the LSOA list.
```

**Comment by TomWFox** [comment]
ready for review?

**Comment by palpatel0** [comment]
> ready for review?

yes, please :)

File: projtools/tests/test_data.py [hunk 3]
```diff
@@ -73,3 +74,12 @@ def test_borough_multi():
...
+    # LSOA is missing Historical Fraud and Forgery and Sexual Offences
+    assert list(np.sort(np.append(data.lsoa.index.get_level_values(3).unique().to_numpy(), ['Sexual Offences', 'Historical Fraud and Forgery']))) == list(np.sort(data.ward.index.get_level_values(2).unique())) #This test checks that the LSOA and ward datasets have the same categories for offences. Two of the missing categories are added to the LSOA list
+    assert list(np.sort(data.borough.index.get_level_values(1).unique())) == list(np.sort(data.ward.index.get_level_values(3).unique())) #This test checks that the borough and ward data sets have the same list of offences
+    # LSOA is missing Rape, Other Sexual Offences and Historical Fraud and Forgery
+    assert list(np.sort(np.append(data.lsoa.index.get_level_values(4).unique().to_numpy(), ['Rape', 'Other Sexual Offences', 'Historical Fraud and Forgery']))) == list(np.sort(data.ward.index.get_level_values(3).unique())) #This test checks that the LSOA and ward datasets have the same list of offences. Three of the missing offences are added to the LSOA list.
```
Comment (TomWFox) [inline projtools/tests/test_data.py:85]:
```suggestion
def test_data_consistency():
    """Tests that the data is the same between the different data sets"""
    # This test checks that the borough and ward data sets have the same categories for offences
    assert list(np.sort(data.borough.index.get_level_values(0).unique())) == list(np.sort(data.ward.index.get_level_values(2).unique()))
    # This test checks that the LSOA and ward datasets have the same categories for offences. Two of the missing categories (Historical Fraud and Forgery and Sexual Offences) are added to the LSOA list
    assert list(np.sort(np.append(data.lsoa.index.get_level_values(3).unique().to_numpy(), ['Sexual Offences', 'Historical Fraud and Forgery']))) == list(np.sort(data.ward.index.get_level_values(2).unique()))
    #This test checks that the borough and ward data sets have the same list of offences
    assert list(np.sort(data.borough.index.get_level_values(1).unique())) == list(np.sort(data.ward.index.get_level_values(3).unique()))
    #This test checks that the LSOA and ward datasets have the same list of offences. Three of the missing offences (Rape, Other Sexual Offences and Historical Fraud and Forgery) are added to the LSOA list.
    assert list(np.sort(np.append(data.lsoa.index.get_level_values(4).unique().to_numpy(), ['Rape', 'Other Sexual Offences', 'Historical Fraud and Forgery']))) == list(np.sort(data.ward.index.get_level_values(3).unique()))
```

**Review by TomWFox — CHANGES_REQUESTED** [review]
Thanks for adding those comments, I think it's better for readability to have them above, rather than at the end of the lines

**Review by TomWFox — APPROVED** [review]

**Pull request merged**